### PR TITLE
Initial commit for Pure Storage Ansible module

### PR DESCRIPTION
**Pull request by sdodsley**
##### SUMMARY
Storage module for Pure Storage FlashArray device. Modules for create/delete/modify volumes, create/delete hosts, attach hosts to volumes, create/modify/delete host groups and create/delete protection groups

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
 - purefa_host
 - purefa_volume
 - purefa_pg
 - purefa_hg

##### ANSIBLE VERSION
```
ansible 2.4.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/pureuser/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/dist-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.12 (default, Nov 19 2016, 06:48:10) [GCC 5.4.0 20160609]
```

### CODE REVIEW

File: lib/ansible/module_utils/pure.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+        module.fail_json(msg="Pure Storage FlashArray authentication failed. Check your credentials")
+    return system
+
+
+def purefa_argument_spec():
+    """Return standard base dictionary used for the argument_spec argument in AnsibleModule"""
+
+    return dict(
+        fa_url=dict(),
+        api_token=dict(),
```
Comment (gundalow) [inline lib/ansible/module_utils/pure.py]: Please add `no_log=true` to ensure the API token gets masked in output

File: lib/ansible/utils/module_docs_fragments/purestorage.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    description:
+      - FlashArray management IPv4 address or Hostname.
+    required: true
+  api_token:
+    description:
+      - FlashArray API token for admin privilaed user.
+    required: true
+notes:
+  - This module requires purestorage python library
+  - You must set PUREFA_URL and PUREFA_API environment variables
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/purestorage.py]: `You must set C(PUREFA_URL) and C(PUREFA_API) environment variables`

File: lib/ansible/utils/module_docs_fragments/purestorage.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    if url and api_token arguments are not passed to the module directly
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/purestorage.py]: `if I(url) and I(api_token) arguments are not passed to the module directly.`

**Review by gundalow — CHANGES_REQUESTED** [review]

**Comment by gundalow** [comment]
@sdodsley Thanks for these new modules, from a quick look they seem sensible.

We've found the following process works best for getting groups of modules integrated
http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html#your-first-pull-request

So if you could remove the other modules from this PR is merged, that would be great.

You may also find some of the documentation & formatting information in http://docs.ansible.com/ansible/dev_guide/developing_modules_in_groups.html useful.

**Comment by sdodsley** [comment]
@gundalow Happy to do these changes, but concerned about timing for the batched PRs after the initial merge. If there is any risk of a RC for 2.4 that would split the PRs then this would make the whole module useless. All modules have to be available to be effective.

**Comment by gundalow** [comment]
If this is the total set of modules, you are fine for 2.4.

Feel free to `@gundalow` in the PRs that get raised after the first one has been merged

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    - This module creates, deletes or modifies hosts on Pure Storage FlashArray.
+author: Simon Dodsley (@simondodsley)
+options:
+  host:
+    description:
+      - Host Name
+    required: true
+  state:
+    description:
+      - Creates/Modifies Host when present or removes when absent
```
Comment (gundalow) [inline lib/ansible/modules/storage/purestorage/purefa_host.py]: This doesn't seem to read right with present vs absent

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+      - When removing all connected volumes will be disconnected
+    required: false
+    default: present
+    choices: [ "present", "absent" ]
+  protocol:
+    description:
+      - Defines the host connection protocol for volumes, either iscsi of fc
```
Comment (gundalow) [inline lib/ansible/modules/storage/purestorage/purefa_host.py]: `Defines the host connection protocol for volumes.`

No need to mention iscsi or fc, docs will be generated by the `choices:` line.

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+description:
+    - This module creates, deletes or modifies hosts on Pure Storage FlashArray.
+author: Simon Dodsley (@simondodsley)
+options:
+  host:
+    description:
+      - Host Name
+    required: true
+  state:
+    description:
```
Comment (gundalow) [inline lib/ansible/modules/storage/purestorage/purefa_host.py:40]: For consistence please ensure all `description:` lines are full sentences (capital letters + full stop at the end)

File: lib/ansible/modules/storage/purestorage/purefa_host.py [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: purefa_host
+version_added: 2.4
```
Comment (gundalow) [inline lib/ansible/modules/storage/purestorage/purefa_host.py]: `version_added: "2.4"`

File: lib/ansible/utils/module_docs_fragments/purestorage.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    # Standard Pure Storage documentation fragment
+    DOCUMENTATION = '''
+options:
+  fa_url:
+    description:
+      - FlashArray management IPv4 address or Hostname.
+    required: true
+  api_token:
+    description:
+      - FlashArray API token for admin privilaed user.
```
Comment (gundalow) [inline lib/ansible/utils/module_docs_fragments/purestorage.py]: privileged

**Review by gundalow — CHANGES_REQUESTED** [review]
Thanks for removing the other modules

**Comment by sdodsley** [comment]
@gundalow Can you have a look to see if these revisions are correct?

**Comment by sdodsley** [comment]
Hi John



Any chance you could have a look at this PR so I can start work on the
others once we have the first one merged?



I modified as per your requests, but one of them isn’t being picked up by
ansibot so it still appears there is a request outstanding.



EDIT by gundalow: deleted tail of original email comment

**Review by gundalow — APPROVED** [review]

**Comment by gundalow** [comment]
Merged, thanks
You can now raise one PR per module for the remaining

**Pull request merged**